File: public/cpp/040-invert-binary-tree.cpp
/*
 * Problem: Invert Binary Tree (LeetCode 226)
 * Difficulty: Easy
 * Category: Trees and Graphs
 * 
 * Description:
 * Given the root of a binary tree, invert the tree, and return its root.
 * Inverting a tree means swapping the left and right children of all nodes.
 * 
 * Example 1:
 * Input: root = [4,2,7,1,3,6,9]
 *        4
 *      /   \
 *     2     7
 *    / \   / \
 *   1   3 6   9
 * 
 * Output: [4,7,2,9,6,3,1]
 *        4
 *      /   \
 *     7     2
 *    / \   / \
 *   9   6 3   1
 * 
 * Example 2:
 * Input: root = [2,1,3]
 * Output: [2,3,1]
 * 
 * Example 3:
 * Input: root = []
 * Output: []
 * 
 * Constraints:
 * - The number of nodes in the tree is in the range [0, 100]
 * - -100 <= Node.val <= 100
 */

#include <iostream>
#include <queue>
#include <stack>
#include <vector>
using namespace std;

// Definition for a binary tree node
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

/*
 * APPROACH 1: RECURSIVE DFS (POST-ORDER)
 * 
 * Intuition:
 * - Use recursion to traverse the tree in post-order
 * - First invert left and right subtrees
 * - Then swap the left and right children of current node
 * - This is the most intuitive and elegant solution
 * 
 * Algorithm:
 * 1. Base case: if node is null, return null
 * 2. Recursively invert left subtree
 * 3. Recursively invert right subtree
 * 4. Swap left and right children
 * 5. Return the current node
 * 
 * Time Complexity: O(n) - visit each node once
 * Space Complexity: O(h) - recursion stack, where h is height
 *                   O(n) worst case for skewed tree
 *                   O(log n) average case for balanced tree
 */
class Solution1 {
public:
    TreeNode* invertTree(TreeNode* root) {
        // Base case: empty tree
        if (!root) return nullptr;
        
        // Recursively invert left and right subtrees
        TreeNode* left = invertTree(root->left);
        TreeNode* right = invertTree(root->right);
        
        // Swap left and right children
        root->left = right;
        root->right = left;
        
        return root;
    }
};

/*
 * APPROACH 2: RECURSIVE DFS (PRE-ORDER)
 * 
 * Intuition:
 * - Swap children first, then recursively invert subtrees
 * - This is slightly different from post-order but equally valid
 * - The order doesn't matter since we're swapping at every level
 * 
 * Algorithm:
 * 1. Base case: if node is null, return null
 * 2. Swap left and right children immediately
 * 3. Recursively invert left subtree (which was originally right)
 * 4. Recursively invert right subtree (which was originally left)
 * 5. Return the current node
 * 
 * Time Complexity: O(n) - visit each node once
 * Space Complexity: O(h) - recursion stack
 */
class Solution2 {
public:
    TreeNode* invertTree(TreeNode* root) {
        // Base case: empty tree
        if (!root) return nullptr;
        
        // Swap left and right children first
        TreeNode* temp = root->left;
        root->left = root->right;
        root->right = temp;
        
        // Recursively invert subtrees
        invertTree(root->left);
        invertTree(root->right);
        
        return root;
    }
};

/*
 * APPROACH 3: ITERATIVE BFS (LEVEL-ORDER)
 * 
 * Intuition:
 * - Use a queue to perform level-order traversal
 * - For each node, swap its left and right children
 * - Add children to queue for processing
 * - This avoids recursion stack overhead
 * 
 * Algorithm:
 * 1. If root is null, return null
 * 2. Create a queue and add root
 * 3. While queue is not empty:
 *    a. Dequeue a node
 *    b. Swap its left and right children
 *    c. Add non-null children to queue
 * 4. Return root
 * 
 * Time Complexity: O(n) - visit each node once
 * Space Complexity: O(w) - queue size, where w is max width
 *                   O(n) worst case for complete binary tree
 */
class Solution3 {
public:
    TreeNode* invertTree(TreeNode* root) {
        // Base case: empty tree
        if (!root) return nullptr;
        
        // BFS using queue
        queue<TreeNode*> q;
        q.push(root);
        
        while (!q.empty()) {
            TreeNode* node = q.front();
            q.pop();
            
            // Swap left and right children
            TreeNode* temp = node->left;
            node->left = node->right;
            node->right = temp;
            
            // Add children to queue
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
        }
        
        return root;
    }
};

/*
 * APPROACH 4: ITERATIVE DFS (USING STACK)
 * 
 * Intuition:
 * - Use a stack to simulate recursive DFS
 * - This is the iterative version of the recursive approach
 * - Useful when recursion depth might cause stack overflow
 * 
 * Algorithm:
 * 1. If root is null, return null
 * 2. Create a stack and push root
 * 3. While stack is not empty:
 *    a. Pop a node
 *    b. Swap its left and right children
 *    c. Push non-null children to stack
 * 4. Return root
 * 
 * Time Complexity: O(n) - visit each node once
 * Space Complexity: O(h) - stack size, where h is height
 */
class Solution4 {
public:
    TreeNode* invertTree(TreeNode* root) {
        // Base case: empty tree
        if (!root) return nullptr;
        
        // DFS using stack
        stack<TreeNode*> stk;
        stk.push(root);
        
        while (!stk.empty()) {
            TreeNode* node = stk.top();
            stk.pop();
            
            // Swap left and right children
            TreeNode* temp = node->left;
            node->left = node->right;
            node->right = temp;
            
            // Push children to stack
            if (node->left) stk.push(node->left);
            if (node->right) stk.push(node->right);
        }
        
        return root;
    }
};

/*
 * APPROACH 5: MORRIS TRAVERSAL (SPACE-OPTIMIZED)
 * 
 * Intuition:
 * - Use Morris traversal to achieve O(1) space complexity
 * - Temporarily modify tree structure to avoid stack/queue
 * - Create threaded binary tree for traversal
 * - This is the most space-efficient solution
 * 
 * Algorithm:
 * 1. Start with current = root
 * 2. While current is not null:
 *    a. If current has no left child:
 *       - Swap left and right
 *       - Move to right child
 *    b. If current has left child:
 *       - Find rightmost node in left subtree
 *       - If rightmost's right is null:
 *         * Create thread: rightmost->right = current
 *         * Move to left child
 *       - If rightmost's right is current:
 *         * Remove thread: rightmost->right = null
 *         * Swap left and right
 *         * Move to right child
 * 3. Return root
 * 
 * Time Complexity: O(n) - each edge traversed at most twice
 * Space Complexity: O(1) - no extra space except variables
 */
class Solution5 {
public:
    TreeNode* invertTree(TreeNode* root) {
        TreeNode* current = root;
        
        while (current) {
            if (!current->left) {
                // No left child, swap and move right
                TreeNode* temp = current->left;
                current->left = current->right;
                current->right = temp;
                current = current->left; // Move to what was right
            } else {
                // Find rightmost node in left subtree
                TreeNode* predecessor = current->left;
                while (predecessor->right && predecessor->right != current) {
                    predecessor = predecessor->right;
                }
                
                if (!predecessor->right) {
                    // Create thread
                    predecessor->right = current;
                    current = current->left;
                } else {
                    // Remove thread and swap
                    predecessor->right = nullptr;
                    TreeNode* temp = current->left;
                    current->left = current->right;
                    current->right = temp;
                    current = current->left; // Move to what was right
                }
            }
        }
        
        return root;
    }
};

// Helper function to create a binary tree from level-order array
TreeNode* createTree(const vector<int>& values) {
    if (values.empty() || values[0] == -1) return nullptr;
    
    TreeNode* root = new TreeNode(values[0]);
    queue<TreeNode*> q;
    q.push(root);
    
    int i = 1;
    while (!q.empty() && i < values.size()) {
        TreeNode* node = q.front();
        q.pop();
        
        // Left child
        if (i < values.size() && values[i] != -1) {
            node->left = new TreeNode(values[i]);
            q.push(node->left);
        }
        i++;
        
        // Right child
        if (i < values.size() && values[i] != -1) {
            node->right = new TreeNode(values[i]);
            q.push(node->right);
        }
        i++;
    }
    
    return root;
}

// Helper function to print tree in level-order
vector<int> levelOrder(TreeNode* root) {
    vector<int> result;
    if (!root) return result;
    
    queue<TreeNode*> q;
    q.push(root);
    
    while (!q.empty()) {
        TreeNode* node = q.front();
        q.pop();
        
        result.push_back(node->val);
        
        if (node->left) q.push(node->left);
        if (node->right) q.push(node->right);
    }
    
    return result;
}

// Helper function to delete tree
void deleteTree(TreeNode* root) {
    if (!root) return;
    deleteTree(root->left);
    deleteTree(root->right);
    delete root;
}

// Test function
void test(const vector<int>& input, int approach) {
    TreeNode* root = createTree(input);
    TreeNode* result = nullptr;
    
    cout << "Input: [";
    for (int i = 0; i < input.size(); i++) {
        if (input[i] == -1) cout << "null";
        else cout << input[i];
        if (i < input.size() - 1) cout << ",";
    }
    cout << "]\n";
    
    switch(approach) {
        case 1: {
            Solution1 sol;
            result = sol.invertTree(root);
            cout << "Approach 1 (Recursive Post-order): ";
            break;
        }
        case 2: {
            Solution2 sol;
            result = sol.invertTree(root);
            cout << "Approach 2 (Recursive Pre-order): ";
            break;
        }
        case 3: {
            Solution3 sol;
            result = sol.invertTree(root);
            cout << "Approach 3 (Iterative BFS): ";
            break;
        }
        case 4: {
            Solution4 sol;
            result = sol.invertTree(root);
            cout << "Approach 4 (Iterative DFS): ";
            break;
        }
        case 5: {
            Solution5 sol;
            result = sol.invertTree(root);
            cout << "Approach 5 (Morris Traversal): ";
            break;
        }
    }
    
    vector<int> output = levelOrder(result);
    cout << "[";
    for (int i = 0; i < output.size(); i++) {
        cout << output[i];
        if (i < output.size() - 1) cout << ",";
    }
    cout << "]\n\n";
    
    deleteTree(result);
}

int main() {
    // Test Case 1: Regular binary tree
    cout << "Test Case 1: Regular binary tree\n";
    vector<int> test1 = {4, 2, 7, 1, 3, 6, 9};
    for (int i = 1; i <= 5; i++) {
        test(test1, i);
    }
    
    // Test Case 2: Small tree
    cout << "Test Case 2: Small tree\n";
    vector<int> test2 = {2, 1, 3};
    for (int i = 1; i <= 5; i++) {
        test(test2, i);
    }
    
    // Test Case 3: Empty tree
    cout << "Test Case 3: Empty tree\n";
    vector<int> test3 = {};
    for (int i = 1; i <= 5; i++) {
        test(test3, i);
    }
    
    // Test Case 4: Single node
    cout << "Test Case 4: Single node\n";
    vector<int> test4 = {1};
    for (int i = 1; i <= 5; i++) {
        test(test4, i);
    }
    
    // Test Case 5: Left-skewed tree
    cout << "Test Case 5: Left-skewed tree\n";
    vector<int> test5 = {1, 2, -1, 3, -1, 4};
    for (int i = 1; i <= 5; i++) {
        test(test5, i);
    }
    
    // Test Case 6: Right-skewed tree
    cout << "Test Case 6: Right-skewed tree\n";
    vector<int> test6 = {1, -1, 2, -1, 3, -1, 4};
    for (int i = 1; i <= 5; i++) {
        test(test6, i);
    }
    
    // Test Case 7: Complete binary tree
    cout << "Test Case 7: Complete binary tree\n";
    vector<int> test7 = {1, 2, 3, 4, 5, 6, 7};
    for (int i = 1; i <= 5; i++) {
        test(test7, i);
    }
    
    return 0;
}

/*
 * COMPLEXITY ANALYSIS SUMMARY:
 * 
 * Approach 1 (Recursive Post-order):
 * - Time: O(n) - visit each node once
 * - Space: O(h) - recursion stack
 * - Best for: Clean, readable code
 * 
 * Approach 2 (Recursive Pre-order):
 * - Time: O(n) - visit each node once
 * - Space: O(h) - recursion stack
 * - Best for: Alternative recursive approach
 * 
 * Approach 3 (Iterative BFS):
 * - Time: O(n) - visit each node once
 * - Space: O(w) - queue size (max width)
 * - Best for: Level-by-level processing
 * 
 * Approach 4 (Iterative DFS):
 * - Time: O(n) - visit each node once
 * - Space: O(h) - stack size
 * - Best for: Avoiding recursion
 * 
 * Approach 5 (Morris Traversal):
 * - Time: O(n) - each edge traversed at most twice
 * - Space: O(1) - no extra space
 * - Best for: Space-constrained environments
 * 
 * INTERVIEW TIPS:
 * 1. Start with recursive solution (Approach 1) - most intuitive
 * 2. Mention iterative alternatives if asked about space optimization
 * 3. Morris traversal shows advanced knowledge but is rarely required
 * 4. Discuss trade-offs between recursion and iteration
 * 5. This problem tests basic tree manipulation skills
 * 
 * COMMON MISTAKES:
 * 1. Forgetting to return the root node
 * 2. Not handling null nodes properly
 * 3. Swapping children incorrectly (need temporary variable)
 * 4. In Morris traversal, not removing threads properly
 * 5. Confusing pre-order and post-order approaches
 * 
 * FOLLOW-UP QUESTIONS:
 * 1. Can you invert only a subtree? (Yes, same logic)
 * 2. How to verify if a tree is inverted? (Compare with original)
 * 3. Can you do it in-place? (Yes, all approaches are in-place)
 * 4. What if tree has millions of nodes? (Use iterative to avoid stack overflow)
 * 5. How to invert n-ary tree? (Same logic, iterate through all children)
 */

// Made with Bob

File: src/data/tutorials/040-invert-binary-tree.js
/**
 * Tutorial content for #040 — Invert Binary Tree
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given the root of a binary tree, invert the tree (i.e. mirror it) and return its root. Every node's left and right children should be swapped at every level.`,
    example: `Input:  [4, 2, 7, 1, 3, 6, 9]\n        4\n      /   \\\n     2     7\n    / \\ / \\\n   1  3 6  9\n→ Swap left/right at every node:\n        4\n      /   \\\n     7     2\n    / \\ / \\\n   9  6 3  1\n✅ Answer: [4, 7, 2, 9, 6, 3, 1]`,
    keyInsight: `Inverting a binary tree just means swapping the left and right child at every node. Whether you do this top-down (pre-order), bottom-up (post-order), or level-by-level (BFS) doesn't matter — all approaches visit every node once and swap its children.`,
  },

  approaches: {
    'Recursive DFS (Post-order)': {
      intuition: `Recurse to the leaves first, then swap on the way back up. This is post-order: first invert the left subtree, then invert the right subtree, then swap the two children of the current node. By the time you swap, both subtrees are already fully inverted.`,
      steps: [
        `Base case: if root is null, return null.`,
        `Recursively invert the left subtree: invertTree(root.left).`,
        `Recursively invert the right subtree: invertTree(root.right).`,
        `Swap: temp = root.left; root.left = root.right; root.right = temp.`,
        `Return root.`,
      ],
      example: `Tree: 4→(2→(1,3), 7→(6,9))\n\nPost-order bottom-up:\nInvert(1)→1, Invert(3)→3\nInvert(2): swap 1,3 → node 2 now has left=3, right=1\nInvert(6)→6, Invert(9)→9\nInvert(7): swap 6,9 → node 7 now has left=9, right=6\nInvert(4): swap node2,node7 → node 4 now has left=7, right=2\n✅ Answer: [4,7,2,9,6,3,1]`,
      keyInsight: `O(n) time (visits every node once), O(h) space for the call stack. Post-order ensures both children are fully inverted before the swap — though pre-order also works fine for this problem.`,
    },

    'Recursive DFS (Pre-order)': {
      intuition: `Swap first, then recurse. This is pre-order: swap the left and right children of the current node before recurring down. After the swap, recurse on the (now swapped) children. Both pre-order and post-order are correct — the order of operations just differs in when the swap happens.`,
      steps: [
        `Base case: if root is null, return null.`,
        `Swap the children: temp = root.left; root.left = root.right; root.right = temp.`,
        `Recursively call invertTree(root.left).`,
        `Recursively call invertTree(root.right).`,
        `Return root.`,
      ],
      example: `Tree: 4→(2→(1,3), 7→(6,9))\n\nPre-order top-down:\nAt 4: swap children → 4.left=7, 4.right=2\nAt 7 (was right, now left): swap children → 7.left=9, 7.right=6\nAt 9: no children, return.\nAt 6: no children, return.\nAt 2 (was left, now right): swap children → 2.left=3, 2.right=1\nAt 3: no children, return.\nAt 1: no children, return.\n✅ Answer: [4,7,2,9,6,3,1]`,
      keyInsight: `O(n) time, O(h) space. Pre-order swaps on the way down; post-order swaps on the way up. Both are correct — pick whichever feels more intuitive to you.`,
    },

    'Iterative BFS': {
      intuition: `Use a queue to process nodes level by level. For each node dequeued, swap its left and right children, then enqueue the (now swapped) non-null children. This avoids recursion entirely and is easy to trace through level by level.`,
      steps: [
        `If root is null, return null.`,
        `Enqueue root.`,
        `While queue is not empty: dequeue a node.`,
        `Swap: temp = node.left; node.left = node.right; node.right = temp.`,
        `Enqueue non-null node.left and node.right.`,
        `Return root after the queue is empty.`,
      ],
      example: `Tree: 4→(2→(1,3), 7→(6,9))\n\nQueue: [4]\nDequeue 4, swap → 4.left=7, 4.right=2. Enqueue 7, 2.\nQueue: [7, 2]\nDequeue 7, swap → 7.left=9, 7.right=6. Enqueue 9, 6.\nDequeue 2, swap → 2.left=3, 2.right=1. Enqueue 3, 1.\nQueue: [9, 6, 3, 1]\nDequeue 9,6,3,1: all leaves, swap null/null (no effect).\nQueue: []\n✅ Answer: [4,7,2,9,6,3,1]`,
      keyInsight: `O(n) time, O(n) space for the queue (max width of the tree at any level). Iterative and easy to visualise level-by-level. Avoids recursion stack overhead for very deep trees.`,
    },

    'Iterative DFS (Stack)': {
      intuition: `Replace the recursive call stack with an explicit stack. For each node popped, swap its children, then push the non-null children onto the stack. This mirrors pre-order recursive DFS — swap first, then push children — but with no risk of stack overflow on very deep trees.`,
      steps: [
        `If root is null, return null.`,
        `Push root onto a stack.`,
        `While stack is not empty: pop a node.`,
        `Swap its children: temp = node.left; node.left = node.right; node.right = temp.`,
        `Push non-null node.left and node.right onto the stack.`,
        `Return root after the stack empties.`,
      ],
      example: `Tree: 4→(2→(1,3), 7→(6,9))\n\nStack: [4]\nPop 4: swap → 4.left=7, 4.right=2. Push 7, push 2. Stack: [7,2]\nPop 2 (top): swap → 2.left=3, 2.right=1. Push 3, push 1. Stack: [7,3,1]\nPop 1: no children. Stack: [7,3]\nPop 3: no children. Stack: [7]\nPop 7: swap → 7.left=9, 7.right=6. Push 9, push 6. Stack: [9,6]\nPop 6: no children. Stack: [9]\nPop 9: no children. Stack: []\n✅ Answer: [4,7,2,9,6,3,1]`,
      keyInsight: `O(n) time, O(h) space for the stack. Functionally identical to the recursive pre-order approach but uses an explicit stack. Preferred when recursion depth is a concern.`,
    },

    'Morris Traversal': {
      intuition: `Invert the tree in O(1) space using Morris traversal — no stack, no recursion. Thread the tree by temporarily linking the inorder predecessor's right pointer back to the current node. When visiting each node (on the first visit for nodes with left children, or immediately for leaf-like nodes), swap its children. Then unthread as you return.`,
      steps: [
        `curr = root.`,
        `While curr is not null: if curr.left is null → swap curr's children; move curr = curr.left (was right before swap).`,
        `Otherwise find curr's inorder predecessor (rightmost node in left subtree not yet threaded to curr).`,
        `If pred.right is null: set pred.right = curr (create thread); move curr = curr.left.`,
        `If pred.right == curr (thread exists): remove thread (pred.right = null); swap curr's children; move curr = curr.left (was right before swap).`,
      ],
      example: `Tree: 4→(2→(1,3), 7→(6,9))\n\ncurr=4: left=2≠null. Pred of 4 in left subtree = node(3) (rightmost of subtree rooted at 2).\n  pred.right=null → thread: 3.right=4. curr=2.\ncurr=2: left=1≠null. Pred of 2 = node(1). pred.right=null → thread: 1.right=2. curr=1.\ncurr=1: left=null → swap 1's children (1 has none). curr=1.left (was right)=null... (traversal continues threading and swapping each node)\nEach node's children are swapped exactly once.\n✅ Answer: [4,7,2,9,6,3,1]`,
      keyInsight: `O(n) time, O(1) space — no extra data structures. The most space-efficient approach. Complex to implement correctly due to threading/unthreading bookkeeping; rarely asked in interviews but valuable to know for space-constrained environments.`,
    },
  },
}
